### DISCUSSION TITLE
How to run on multi-GPU platforms

### DISCUSSION THREAD

**Discussion by JiahuaZhao** [ExtremeFLOW/neko#1761] [Q&A]
Hi, I compiled Neko (#540) with openmpi (--with-cuda) and when I use `mpirun --np 2 . /neko tgv_Re360.case` with the following error:
![Snipaste_2024-04-01_17-06-35](https://github.com/ExtremeFLOW/neko/assets/48776996/974f78be-36f1-4b14-bafa-f6dd05ed867f)
Also I found out that all MPI processes are running on one GPU.

So how to run the multi-GPU version correctly?

Thanks!

**Comment by njansson** [comment]
Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh

**Comment by JiahuaZhao** [comment]
> Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh

Thanks! I solved the multi-GPU resource allocation issue based on your suggestion, but currently `an illegal memory access was encountered` is still not solved. Do you have any advice on this issue? I am using two V100-16GB GPUs with CUDA version 12.4.

**Comment by njansson** [comment]
> > Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh
> 
> Thanks! I solved the multi-GPU resource allocation issue based on your suggestion, but currently `an illegal memory access was encountered` is still not solved. Do you have any advice on this issue? I am using two V100-16GB GPUs with CUDA version 12.4.

I would assume the error is due to a too small case, `tgv_Re360.case` I think only has 512 elements. One should strive to have roughly 5k - 15k elements per device. 

(Ps. errors for small cases are targeted to be fixed in v0.8 of Neko)

**Comment by JiahuaZhao** [comment]
> > > Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh
> > 
> > 
> > Thanks! I solved the multi-GPU resource allocation issue based on your suggestion, but currently `an illegal memory access was encountered` is still not solved. Do you have any advice on this issue? I am using two V100-16GB GPUs with CUDA version 12.4.
> 
> I would assume the error is due to a too small case, `tgv_Re360.case` I think only has 512 elements. One should strive to have roughly 5k - 15k elements per device.
> 
> (Ps. errors for small cases are targeted to be fixed in v0.8 of Neko)

Thanks I am using a larger case (32768.nmsh), and using 4 GPUs at the same time (each GPU is evenly distributed to 8K elements), but I am still getting the following error... 
![Snipaste_2024-04-02_15-16-34](https://github.com/ExtremeFLOW/neko/assets/48776996/96af80bd-91a3-4ea9-b5ce-d389cf74e49f)
is there something wrong with the way I am running it?

**Comment by njansson** [comment]
> > > > Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh
> > > 
> > > 
> > > Thanks! I solved the multi-GPU resource allocation issue based on your suggestion, but currently `an illegal memory access was encountered` is still not solved. Do you have any advice on this issue? I am using two V100-16GB GPUs with CUDA version 12.4.
> > 
> > 
> > I would assume the error is due to a too small case, `tgv_Re360.case` I think only has 512 elements. One should strive to have roughly 5k - 15k elements per device.
> > (Ps. errors for small cases are targeted to be fixed in v0.8 of Neko)
> 
> Thanks I am using a larger case (32768.nmsh), and using 4 GPUs at the same time (each GPU is evenly distributed to 8K elements), but I am still getting the following error... ![Snipaste_2024-04-02_15-16-34](https://private-user-images.githubusercontent.com/48776996/318808053-96af80bd-91a3-4ea9-b5ce-d389cf74e49f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNjY4MDMsIm5iZiI6MTcxMjA2NjUwMywicGF0aCI6Ii80ODc3Njk5Ni8zMTg4MDgwNTMtOTZhZjgwYmQtOTFhMy00ZWE5LWI1Y2UtZDM4OWNmNzRlNDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE0MDE0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0YjM5YzY0ZjZlYzY4MzlmMjdmOWFkMDA4ODNkNjYyZWQ5OWUzMDY5YjBkMjQwMjRkNTYwMWI0MjE5ZmQ2YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gF3cOMiC61Ak-aQK0UercrLrrTn-LrJ2QhHSwFQ3GDs) is there something wrong with the way I am running it?

For which version of Neko is this?

**Comment by JiahuaZhao** [comment]
> > > > > Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh
> > > > 
> > > > 
> > > > Thanks! I solved the multi-GPU resource allocation issue based on your suggestion, but currently `an illegal memory access was encountered` is still not solved. Do you have any advice on this issue? I am using two V100-16GB GPUs with CUDA version 12.4.
> > > 
> > > 
> > > I would assume the error is due to a too small case, `tgv_Re360.case` I think only has 512 elements. One should strive to have roughly 5k - 15k elements per device.
> > > (Ps. errors for small cases are targeted to be fixed in v0.8 of Neko)
> > 
> > 
> > Thanks I am using a larger case (32768.nmsh), and using 4 GPUs at the same time (each GPU is evenly distributed to 8K elements), but I am still getting the following error... ![Snipaste_2024-04-02_15-16-34](https://private-user-images.githubusercontent.com/48776996/318808053-96af80bd-91a3-4ea9-b5ce-d389cf74e49f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNjY4MDMsIm5iZiI6MTcxMjA2NjUwMywicGF0aCI6Ii80ODc3Njk5Ni8zMTg4MDgwNTMtOTZhZjgwYmQtOTFhMy00ZWE5LWI1Y2UtZDM4OWNmNzRlNDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE0MDE0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0YjM5YzY0ZjZlYzY4MzlmMjdmOWFkMDA4ODNkNjYyZWQ5OWUzMDY5YjBkMjQwMjRkNTYwMWI0MjE5ZmQ2YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gF3cOMiC61Ak-aQK0UercrLrrTn-LrJ2QhHSwFQ3GDs) is there something wrong with the way I am running it?
> 
> For which version of Neko is this?

v0.6.1

**Comment by njansson** [comment]
> > > > > > Neko assumes one device per mpi rank, and this has to be assigned from the environment, typically via environment variables or a wrapper script, see for example https://github.com/ExtremeFLOW/flettner_rotor/blob/main/run.sh
> > > > > 
> > > > > 
> > > > > Thanks! I solved the multi-GPU resource allocation issue based on your suggestion, but currently `an illegal memory access was encountered` is still not solved. Do you have any advice on this issue? I am using two V100-16GB GPUs with CUDA version 12.4.
> > > > 
> > > > 
> > > > I would assume the error is due to a too small case, `tgv_Re360.case` I think only has 512 elements. One should strive to have roughly 5k - 15k elements per device.
> > > > (Ps. errors for small cases are targeted to be fixed in v0.8 of Neko)
> > > 
> > > 
> > > Thanks I am using a larger case (32768.nmsh), and using 4 GPUs at the same time (each GPU is evenly distributed to 8K elements), but I am still getting the following error... ![Snipaste_2024-04-02_15-16-34](https://private-user-images.githubusercontent.com/48776996/318808053-96af80bd-91a3-4ea9-b5ce-d389cf74e49f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNjY4MDMsIm5iZiI6MTcxMjA2NjUwMywicGF0aCI6Ii80ODc3Njk5Ni8zMTg4MDgwNTMtOTZhZjgwYmQtOTFhMy00ZWE5LWI1Y2UtZDM4OWNmNzRlNDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAyVDE0MDE0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0YjM5YzY0ZjZlYzY4MzlmMjdmOWFkMDA4ODNkNjYyZWQ5OWUzMDY5YjBkMjQwMjRkNTYwMWI0MjE5ZmQ2YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gF3cOMiC61Ak-aQK0UercrLrrTn-LrJ2QhHSwFQ3GDs) is there something wrong with the way I am running it?
> > 
> > 
> > For which version of Neko is this?
> 
> v0.6.1

To better understand what is happening, please share how Neko was built (configure flags), and also a log of the failed run (from start to fault)

**Comment by JiahuaZhao** [comment]
@njansson **here are the commands to compile:**
`./configure --prefix=/path/neko-0.6.1 --with-cuda=/path/cuda/v12.4.0 CUDA_CFLAGS=-O3  CUDA_ARCH=-arch=sm_70 NVCC=/path/v12.4.0/bin/nvcc --enable-device-mpi`
'make && make install'

**and outputs:**
  _  __  ____  __ __  ____
   / |/ / / __/ / //_/ / __ \
  /    / / _/  / ,<   / /_/ /
 /_/|_/ /___/ /_/|_|  \____/

 (version: 0.6.1)
 (build: 2024-04-02 on x86_64-pc-linux-gnu using gnu)


 -------Job Information--------
 Start time: 09:50 / 2024-04-02
 Running on: 4 MPI ranks
 CPU type  : Intel(R) Xeon(R) Platinum 8168 CPU @ 2.70GHz
 Bcknd type: Accelerator (CUDA)
 Dev. name : Tesla V100-SXM3-32GB
 Real type : double precision

 -------------Case-------------
 Reading case file tgv_Re1600.case

   -------------Mesh-------------
   Reading a binary Neko file 32768.nmsh
   gdim = 3, nelements =    32768
   Reading elements
   Reading BC/zone data
   Reading deformation data
   Mesh read, setting up connectivity
   Done setting up mesh and connectivity

   ------------Fluid-------------
   Type       : Modular (Pn/Pn)
   lx         : 8
   Re         : 1.600000E+03
   rho        : 1.000000E+00
   mu         : 1.000000E+00
   Ksp vel.   : (cg, jacobi)
    `-abs tol : 1.000000E-09
   Ksp prs.   : (gmres, hsmg)
    `-abs tol : 1.000000E-09
   Dealias    : T
   Save bdry  : F

      --------Gather-Scatter--------
      Comm         :   Device MPI
      Avg. internal:      2293760
      Avg. external:       131072
      Backend      :         cuda
      Avg. strtgy  :         [10]

      --------Gather-Scatter--------
      Comm         :   Device MPI
      Avg. internal:        57344
      Avg. external:         8192
      Backend      :         cuda
      Avg. strtgy  :         [01]

      --------Gather-Scatter--------
      Comm         :   Device MPI
      Avg. internal:       425984
      Avg. external:        32768
      Backend      :         cuda
      Avg. strtgy  :         [10]

      --------Gather-Scatter--------
      Comm         :   Device MPI
      Avg. internal:      3792896
      Avg. external:       204800
      Backend      :         cuda
      Avg. strtgy  :         [10]

      --------Gather-Scatter--------
      Comm         :   Device MPI
      Avg. internal:      1171456
      Avg. external:        73728
      Backend      :         cuda
      Avg. strtgy  :         [00]

   -----Adding write output------
   File name: field.fld
   Write control: nsamples
   Total samples:           100
   Writes per time unit (Freq.):  5.000000E+01
   Time between writes:  2.000000E-02

 -----Starting simulation------
 T  : [  0.0000000E+00,  0.2000000E+01)
 dt :    0.5000000E-03

   --------Postprocessing--------

      --------Writer output---------
      File name: field.fld
      Output number:     0
      Writing at time:   0.000000 Output time (s):  3.157927

      ---Autotune dudxyz (lx: 8)----
      Chose      : 2 (KSTEP)
POST: t:   0.000000000E+00 Ekin:   0.125000000E+00 enst:   0.375000000E+00

[LOG_CAT_GPU] Cuda failure gpu_cuda.c:137 'an illegal memory access was encountered'
[LOG_CAT_GPU] Cuda failure gpu_cuda.c:137 'an illegal memory access was encountered'
[LOG_CAT_GPU] Cuda failure gpu_cuda.c:137 'an illegal memory access was encountered'
[LOG_CAT_GPU] Cuda failure gpu_cuda.c:137 'an illegal memory access was encountered'
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[17521,1],1]
  Exit code:    1
--------------------------------------------------------------------------

**Comment by njansson** [comment]
As a start I would suggest to first verify that multi gpu run works by disabling device aware MPI (don't add `--enable-device-mpi` to a clean build), remember to always do `make clean` when changing options that affects compute backends.

**Comment by JiahuaZhao** [comment]
> As a start I would suggest to first verify that multi gpu run works by disabling device aware MPI (don't add to a clean build), remember to always do when changing options that affects compute backends.`--enable-device-mpi``make clean`

It worked, thanks! I also wanted to ask though, my openmpi supports cuda, why can't I use `--enable-device-mpi`, or why running neko after using it gives me the error before?
